File: 301_bfs.java
class Solution {
    public List<String> removeInvalidParentheses(String s) {
        List<String> res = new ArrayList<>();
        int max = 0;
        Set<String> checked = new HashSet<>();
        Queue<String> q = new LinkedList<>();
        boolean found = false;
        q.offer(s);
        while (!q.isEmpty()) {
            String cur = q.poll();
            if (isValid(cur)) {
                found = true;
                if (cur.length() >= max && !res.contains(cur)) {
                    max = cur.length();
                    res.add(cur);
                }
            }
            if (found) continue;
            for (int i = 0; i < cur.length(); i++) {
                String sub1 = cur.substring(0, i);
                String sub2 = cur.substring(i + 1, cur.length());
                String next = sub1.concat(sub2);
                if (!checked.contains(next)) {
                    q.offer(next);
                    checked.add(next);
                }
            }
        }
        return res;
        
    }
    
    private boolean isValid(String s) {
        int count = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') count++;
            else if (s.charAt(i) == ')') {
                if (count == 0) return false;
                count--;
            }
        }
        return count == 0;
    }
}

File: 301_bfs.cpp
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        queue<string> cur, next;
        cur.push(s);
        unordered_set<string> res;
        unordered_set<string> visited;
        visited.insert(s);
        while(!cur.empty()) {
            while (!cur.empty()) {
                string node = cur.front(); cur.pop();
                if (isValid(node)) {
                    res.insert(node);
                    continue;
                }

                for (int i = 0; i < node.size(); i++) {
                    if (node[i] == '(' || node[i] == ')') {
                        string tmp = node;
                        tmp.erase(i, 1);
                        if (visited.find(tmp) != visited.end()) {
                            continue;
                        }
                        next.push(tmp);
                        visited.insert(tmp);
                    }
                }
            }
            if (res.size() > 0) {
                break;
            }
            swap(cur, next);
        }
        return vector<string>(res.begin(), res.end());
    }

    bool isValid(string& s) {
        int cnt = 0;
        for (auto c : s) {
            if (c == '(') {
                ++cnt;
            }
            if (c == ')') {
                --cnt;
            }
            if (cnt < 0) {
                return false;
            }
        }
        return cnt == 0;
    }
};

